function_default_parameter_at_end false positive

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Running SwiftLint on the following function results in a false positive for the `function_default_parameter_at_end` rule:

```swift
static func alertWithSingleAction(buttonTitle: String, title: String? = nil, message: String? = nil, buttonHandler: @escaping AlertActionHandler = { _ in }) -> UIAlertController {
    let controller = UIAlertController(title: title, message: message, preferredStyle: .alert)
    let defaultAction = UIAlertAction(title: buttonTitle, style: .default, handler: buttonHandler)
    controller.addAction(defaultAction)
    return controller
}
```

Strangely, other functions with default closures at the end work just fine. Perhaps the multiple default parameters throw it off?

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
/Users/jshier/Desktop/Code/project/Project/source/UIAlertController.swift:32:5: warning: Function Default Parameter at End Violation: Prefer to locate parameters with defaults toward the end of the parameter list. (function_default_parameter_at_end)
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? **0.33.0**
* Installation method used (Homebrew, CocoaPods, building from source, etc)? **Homebrew**
* Paste your configuration file:

```yml
included:
  - Project
  - ProjectTests
  - Project UI Tests
  - DesignableUI
excluded:
  - Pods
disabled_rules:
  - file_length
  - force_cast
  - force_try
  - function_body_length
  - function_parameter_count
  - identifier_name # Configure minimum and maximum length.
  - line_length
  - nesting
  - sorted_imports # SwiftFormat will automatically sort them for us.
  - todo
  - trailing_whitespace
  - type_body_length
opt_in_rules:
  - anyobject_protocol
  - array_init
  - attributes
  - block_based_kvo
  - closure_end_indentation
  - closure_spacing
  - collection_alignment
  - contains_over_first_not_nil
  - convenience_type
  - discouraged_object_literal
  - empty_count
  - empty_string
  - empty_xctest_method
  - explicit_init
  - extension_access_modifier
  - fallthrough
  - fatal_error_message
  - first_where
  - for_where
  - function_default_parameter_at_end
  - identical_operands
  - implicit_return
  - joined_default_parameter
  - legacy_random
  - let_var_whitespace
  - literal_expression_end_indentation
  - lower_acl_than_parent
  - modifier_order
  - multiline_parameters
  - operator_usage_whitespace
  - overridden_super_call
  - pattern_matching_keywords
  - prohibited_super_call
  - redundant_nil_coalescing
  - sorted_first_last
  - sorted_imports
  - strong_iboutlet
  - toggle_bool
  - trailing_closure
  - unused_import
  - unused_private_declaration
  - vertical_parameter_alignment_on_call
  
modifier_order:
  preferred_modifier_order: [acl, override]
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **No.**
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcode-select -p`)? **10.2.1**
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
static func alertWithSingleAction(buttonTitle: String, title: String? = nil, message: String? = nil, buttonHandler: @escaping AlertActionHandler = { _ in }) -> UIAlertController {
    let controller = UIAlertController(title: title, message: message, preferredStyle: .alert)
    let defaultAction = UIAlertAction(title: buttonTitle, style: .default, handler: buttonHandler)
    controller.addAction(defaultAction)
    return controller
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#2792](https://github.com/realm/SwiftLint/issues/2792)
 
+* Fix false positive in `function_default_parameter_at_end` rule when using
+  a closure parameter with default value.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#2788](https://github.com/realm/SwiftLint/issues/2788)
+
 ## 0.33.0: Worldwide Dryers Conference
 
 #### Breaking

--- a/Rules.md
+++ b/Rules.md
@@ -8399,6 +8399,11 @@ func foo(a: Int, b: CGFloat = 0) {
 }
 ```
 
+```swift
+func foo(a: String, b: String? = nil,
+         c: String? = nil, d: @escaping AlertActionHandler = { _ in }) {}
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>

--- a/Source/SwiftLintFramework/Rules/Idiomatic/FunctionDefaultParameterAtEndRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/FunctionDefaultParameterAtEndRule.swift
@@ -26,6 +26,10 @@ public struct FunctionDefaultParameterAtEndRule: ASTRule, ConfigurationProviderR
             func foo(a: Int, b: CGFloat = 0) {
               let block = { (error: Error?) in }
             }
+            """,
+            """
+            func foo(a: String, b: String? = nil,
+                     c: String? = nil, d: @escaping AlertActionHandler = { _ in }) {}
             """
         ],
         triggeringExamples: [
@@ -43,13 +47,22 @@ public struct FunctionDefaultParameterAtEndRule: ASTRule, ConfigurationProviderR
         }
 
         let isNotClosure = { !self.isClosureParameter(dictionary: $0) }
-        let params = dictionary.enclosedVarParameters.filter(isNotClosure).filter { param in
-            guard let paramOffset = param.offset else {
-                return false
-            }
+        let params = dictionary.substructure
+            .flatMap { subDict -> [[String: SourceKitRepresentable]] in
+                guard subDict.kind.flatMap(SwiftDeclarationKind.init) == .varParameter else {
+                    return []
+                }
 
-            return paramOffset < bodyOffset
-        }
+                return [subDict]
+            }
+            .filter(isNotClosure)
+            .filter { param in
+                guard let paramOffset = param.offset else {
+                    return false
+                }
+
+                return paramOffset < bodyOffset
+            }
 
         guard !params.isEmpty else {
             return []

